### ISSUE TITLE
Compare computed keys by tokens in no-dupe-keys

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12070]
**What rule do you want to change?**

no-dupe-keys

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

I'm not sure, it could be new default behavior.

**Please provide some example code that this change will affect:**

```js
/* eslint no-dupe-keys:error */

var foo = {
  [a]: 1,
  [a]: 2
}
```
**What does the rule currently do for this code?**

No warnings.

**What will the rule do after it's changed?**

Warn about the duplicate key.

**Are you willing to submit a pull request to implement this change?**

Yes, I'll be glad to do it.

**Comment by platinumazure** [comment]
What should happen in this situation, where the keys have side effects?

```js
/* eslint no-dupe-keys:error */

var foo = {
  [a++]: 1,
  [a++]: 2,

  [b()]: 3,
  [b()]: 4
}
```

**Comment by mdjermanovic** [comment]
That would be bad. Maybe to compare just identifiers, I guess that's the most common use case.

Albeit, in theory there could be  something like this:

```js
var foo = {
  [a]: a++,
  [a]: 2,
}
```

**Comment by kaicataldo** [comment]
I know [my comment](https://github.com/eslint/eslint/pull/12069#discussion_r311306518) kicked off this discussion, but given the discussion above, I agree that this doesn't seem like a case we can enforce reliably and safely. I think the current behavior probably makes the most sense. @mdjermanovic Thanks for exploring this!

**Comment by dwelle** [comment]
IMO this is a useful feature to implement, even if for the simple case of:

```js
var foo = {
  [a]: 1,
  [a]: 2,
}
```
where the key isn't an expression (as @platinumazure noted) other than a simple variable reference.

In our team, we've already had at least one painful debugging experience due to this.

If the worry is that doing e.g. `{ [o.id]: "a", [o.id]: "b" }` could result in false positive in case `o.id` was a self-incrementing getter (how common is that?), then we can add a rule exception for computed keys if one desires, or don't lint for nested computed keys.

Or is there something else I'm not seeing here?
